### DISCUSSION TITLE
synthetic Antenna logic

### DISCUSSION THREAD

**Discussion by TonyChahoud** [NVlabs/sionna#491] [General]
Good morning Sir, I hope you're doing well.

I would like to request assistance or clarification on the following issue. I am working with Sionna on a real indoor scene, and everything is going well. However, I need some help understanding the logic behind the synthetic array (when it is set to True). I am dealing with 16 antennas on each transmitter in my scene. With synthetic arrays, we end up with the path of the synthetic antenna and not each individual one. This is critical for me because I am interested in the angles of departure and arrival, which are extracted simply. However, I want to relate the power to each path of the synthetic antenna in addition to the four angles.

I have not found a way to extract the channel gain (a, tau) for the synthetic paths. What I used to do was get them for the 16 antennas of that transmitter and then sum them, as I understood from the tutorial that the synthetic array actually emulates all antenna paths. So, assuming that each of the 16 antennas has 30 paths, by synthetic logic, we end up with 1 synthetic antenna with 30 paths. I calculate the channel gain of those 30 new paths through summation of the old ones, using this logic to get the path coefficient for each path (complex number) and from it, I calculate the power.

Is the correct approach to sum, average, or take the maximum of these path coefficients/gains? Can you please clarify if my logic is correct or if there is any other approach to extract the power directly of these synthetic paths?

Thank you.

**Comment by jhoydis** [comment]
Hi @TonyChahoud,

Would you have a code snippet showing what you are doing (using one of the scenes that are already shipped with Sionna)?

**Comment by lllck** [comment]
Hello, thank you for your amazing work.
I also have some questions about synthetic antenna logic and how to calculate power.

1. When I set up both the transmitting and receiving antenna arrays to have only one antenna, I used two methods to calculate power. 
```
import os
import numpy as np
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'
from sionna.rt import load_scene, Transmitter, Receiver, PlanarArray

def _load_scene(scene):
    my_object = scene.get("qingdao_buildings")
    if not my_object == None:
        my_object.radio_material = 'itu_concrete'
    my_object = scene.get("qingdao_ground")
    if not my_object == None:
        my_object.radio_material = "itu_very_dry_ground"
    
    # Configure antenna array for all transmitters
    scene.tx_array = PlanarArray(num_rows=1,
                                 num_cols=1,
                                 vertical_spacing=0.5,
                                 horizontal_spacing=0.5,
                                 pattern='dipole',
                                 polarization="V")

    # Configure antenna array for all receivers
    scene.rx_array = PlanarArray(num_rows=1,
                                 num_cols=1,
                                 vertical_spacing=0.5,
                                 horizontal_spacing=0.5,
                                 pattern='dipole',
                                 polarization="V")
    scene.synthetic_array = True

    return scene

def calculate_for_point_to_point(tx_x,tx_y,tx_z,rx_x,rx_y,rx_z,f, file_path = '/data/qingdao.xml'):
    xml_path = file_path
    scene = load_scene(xml_path)
    scene.frequency = f
    tx = Transmitter(name="tx",position=[tx_x,tx_y,tx_z])
    scene.add(tx)
    rx = Receiver(name="rx",position = [rx_x,rx_y,rx_z])
    scene.add(rx)
    # print('tx and rx have been added!')
    scene = _load_scene(scene)
    paths = scene.compute_paths(max_depth=3, num_samples=1e6, los=True, reflection=True, scattering=False,
                                                diffraction=True, edge_diffraction=True, check_scene=True)
    paths.export('paths.obj')

    a = np.squeeze(paths.a)
    tau = np.squeeze(paths.tau)
    theta_r = np.squeeze(paths.theta_r)
    phi_r = np.squeeze(paths.phi_r)
    mask = np.squeeze(paths.mask)

    power1 = 20 * np.log10(abs(np.sum(paths.a)))
    power2 = _cal_power(a, tau, theta_r, phi_r, mask, f)

    print('Power1:', power1)
    print('Power2:', power2)
    
def _cal_power(a, tau, theta_r, phi_r, mask ,f):
    from math import sin, cos, pi
    x, y, z = 0, 0, 0
    try:
        for i in range(len(a)):
            if mask[i]:
                phi = 2 * pi * f * tau[i]
                tmp = a[i] * (cos(phi) + sin(phi) * 1j)
                # tmp = np.abs(tmp)
                x = x + tmp * sin(theta_r[i]) * cos(phi_r[i])
                y = y + tmp * sin(theta_r[i]) * sin(phi_r[i])
                z = z + tmp * cos(theta_r[i])
    except Exception as e:
        print(e)
        return -np.inf
    
    res = np.sqrt(x ** 2 + y ** 2 + z ** 2)
    np.seterr(divide='ignore')
    return 20 * np.log10(np.abs(res))

def main():
    calculate_for_point_to_point(tx_x = -161, tx_y = -85, tx_z = 70, rx_x= -712 , rx_y = -667, rx_z = 30, f=2.8e9)

if __name__ == '__main__':
    main()

```
The printed result is
```
Power1: -108.83469581604004
Power2: -105.49345034585984
```
Which of power1 and power2 is correct?

2. When I set the transmitting and receiving antenna arrays to have multiple antennas and `synthentic_array=True`, just like 
```
 # Configure antenna array for all transmitters
    scene.tx_array = PlanarArray(num_rows=2,
                                 num_cols=2,
                                 vertical_spacing=0.5,
                                 horizontal_spacing=0.5,
                                 pattern='dipole',
                                 polarization="V")

    # Configure antenna array for all receivers
    scene.rx_array = PlanarArray(num_rows=2,
                                 num_cols=2,
                                 vertical_spacing=0.5,
                                 horizontal_spacing=0.5,
                                 pattern='dipole',
                                 polarization="V")
    scene.synthetic_array = True
```
the dimension of `paths.a` is [1, 1, 4, 1, 4, 14, 1], while the dimensions of `paths.tau`,`paths.theta_r`,`paths.phi_r` and `paths.mask` are [1, 1, 1, 14]. In this case, how should I calculate the power at the receiver?

**Reply by SebastianCa** [reply]
Hi @lllck,
I’m not entirely sure I understand the intent behind your second method, but the first one looks ok. For more details, please refer to the [Ray Tracing Tutorial](https://nvlabs.github.io/sionna/examples/Sionna_Ray_Tracing_Diffraction.html#Experiments-with-a-Simple-Wedge), particularly the `compute_gains()` function.

If you enable the `synthetic_array`, whether you compute the received power over all antennas or per antenna depends on your specific requirements. For more information on the shapes involved, please see issue #547.
